### PR TITLE
cmd: wait for the cluster to be initialized

### PR DESCRIPTION
**Pull request by staebler**
After creating the cluster, wait until the ClusterVersion object indicates that the cluster has been initialized prior to exiting the installer.

### CODE REVIEW

**Comment by staebler** [comment]
This is the error from the installer when run with these changes.

`FATAL Cluster failed to initialize: Could not update clusteroperator "cluster-monitoring-operator" (config.openshift.io/v1, 177 of 241)`

**Comment by wking** [comment]
> But the cluster did eventually initialize...

Would be interesting to see how it recovered.  Maybe it will turn up in CI where we'll have logs.

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -278,7 +284,67 @@ func destroyBootstrap(ctx context.Context, config *rest.Config, directory string
...
+			if silenceRemaining > 0 {
+				return
+			}
+		}
+		logrus.Debugf(msg)
+		silenceRemaining = logDownsample
+		lastLoggedMsg = msg
+	}
+	var failureErr error
+	wait.Until(func() {
```
Comment (smarterclayton) [inline cmd/openshift-install/create.go]: You should be able to use watch.UntilSync now

File: cmd/openshift-install/create.go [hunk 2]
```diff
@@ -278,7 +284,67 @@ func destroyBootstrap(ctx context.Context, config *rest.Config, directory string
...
+		cv, err := cc.ConfigV1().ClusterVersions().Get("version", metav1.GetOptions{})
+		if err != nil {
+			logMessage(fmt.Sprintf("Still waiting for the ClusterVersion: %v", err))
+			return
+		}
+		for _, cond := range cv.Status.Conditions {
```
Comment (smarterclayton) [inline cmd/openshift-install/create.go]: Pull this out into a “findStatusCondition” instead, makes it much easier to read and reason about.

Comment (abhinavdahiya) [reply cmd/openshift-install/create.go]: similar to https://github.com/openshift/library-go/blob/86f3eb4ba0e678146cf47f61010c34c638cea2d2/pkg/config/clusteroperator/v1helpers/status.go#L53-L62

**Comment by abhinavdahiya** [comment]
/approve

This looks good, will wait for sometime to allow other people to comment.
@smarterclayton @wking PTAL

File: cmd/openshift-install/create.go [hunk 3]
```diff
@@ -107,6 +113,10 @@ var (
 					logrus.Fatal(err)
 				}
 
+				if err := waitForInitializedCluster(ctx, config); err != nil {
+					logrus.Fatal(err)
+				}
+
 				consoleURL, err := waitForConsole(ctx, config, rootOpts.dir)
```
Comment (wking) [inline cmd/openshift-install/create.go]: Can we drop this now?  Or can `waitForInitializedCluster` complete before the console route exists?

Comment (abhinavdahiya) [reply cmd/openshift-install/create.go]: installer being a little resilient when fetching the route doesn't seem too off-the-top. so IMO we can keep the fetch/wait for console route

File: cmd/openshift-install/create.go [hunk 4]
```diff
@@ -278,7 +288,50 @@ func destroyBootstrap(ctx context.Context, config *rest.Config, directory string
...
+				cv, ok := event.Object.(*configv1.ClusterVersion)
+				if !ok {
+					logrus.Warnf("Expected a ClusterVersion object but got a %q object instead", event.Object.GetObjectKind().GroupVersionKind())
+					return false, nil
+				}
+				if cov1helpers.IsStatusConditionTrue(cv.Status.Conditions, configv1.OperatorAvailable) {
+					logrus.Debug("Cluster is initialized")
+					return true, nil
+				}
+				if cov1helpers.IsStatusConditionTrue(cv.Status.Conditions, configv1.OperatorFailing) {
```
Comment (wking) [inline cmd/openshift-install/create.go]: Should we expand this to accept `OperatorProgressing` too?  Or does that not have useful messages?

**Comment by smarterclayton** [comment]
/test e2e-aws

Sorry, using you as a guinea pig

File: cmd/openshift-install/create.go (hunk 3 above)
Comment (staebler) [reply cmd/openshift-install/create.go]: I tried removing the wait--thinking as well that it was no longer necessary. But the console route was still not up at the cluster version became available.

File: cmd/openshift-install/create.go (hunk 4 above)
Comment (staebler) [reply cmd/openshift-install/create.go]: There does not appear to have anything more useful in the `Progressing` condition. These are the messages that I have seen in that condition.

When `Failing` is `False`,
```
Working towards 4.0.0-0.alpha-2019-01-30-125030
```

When `Failing` is `True`,
```
Unable to apply 4.0.0-0.alpha-2019-01-30-125030: the update could not be applied
```

**Comment by wking** [comment]
[e2e-aws][1]:

```
Flaky tests:

[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Dynamic PV (default fs)] subPath should support existing directory [Suite:openshift/conformance/parallel] [Suite:k8s]

Failing tests:

[sig-apps] StatefulSet [k8s.io] Basic StatefulSet functionality [StatefulSetBasic] should not deadlock when a pod's predecessor fails [Suite:openshift/conformance/parallel] [Suite:k8s]
```

Notes on the StatefulSet panics [here][2].

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1132/pull-ci-openshift-installer-master-e2e-aws/3268/build-log.txt
[2]: https://github.com/openshift/installer/pull/1155#issuecomment-459212975

**Pull request merged**